libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,294 @@
+    const float distance_to_dest = current_loc.get_distance(destination);
+
+    // lookahead distance is adjusted dynamically based on avoidance results
+    _current_lookahead = constrain_float(_current_lookahead, _lookahead * 0.5f, _lookahead);
+
+    // calculate lookahead dist and time for step1.  distance can be slightly longer than
+    // the distance to the destination to allow room to dodge after reaching the destination
+    const float lookahead_step1_dist = MIN(_current_lookahead, distance_to_dest + OA_BENDYRULER_LOOKAHEAD_PAST_DEST);
+
+    // calculate lookahead dist for step2
+    const float lookahead_step2_dist = _current_lookahead * OA_BENDYRULER_LOOKAHEAD_STEP2_RATIO;
+
+    // get ground course
+    float ground_course_deg;
+    if (ground_speed_vec.length_squared() < OA_BENDYRULER_LOW_SPEED_SQUARED) {
+        // with zero ground speed use vehicle's heading
+        ground_course_deg = AP::ahrs().yaw_sensor * 0.01f;
+    } else {
+        ground_course_deg = degrees(ground_speed_vec.angle());
+    }
+
+    // check OA_BEARING_INC definition allows checking in all directions
+    static_assert(360 % OA_BENDYRULER_BEARING_INC == 0, "check 360 is a multiple of OA_BEARING_INC");
+
+    // search in 5 degree increments around the vehicle alternating left
+    // and right. For each direction check if vehicle would avoid all obstacles
+    float best_bearing = bearing_to_dest;
+    bool have_best_bearing = false;
+    float best_margin = -FLT_MAX;
+    float best_margin_bearing = best_bearing;
+
+    for (uint16_t i = 0; i < (360 / OA_BENDYRULER_BEARING_INC); i++) {
+        float bearing_delta = i * OA_BENDYRULER_BEARING_INC * 0.5f;
+        if (i & 1) {
+            // we alternate between trying to the left and right of the target
+            bearing_delta = -bearing_delta;
+        }
+        // bearing that we are probing
+        const float bearing_test = wrap_180(bearing_to_dest + bearing_delta);
+        const float bearing_change_deg = wrap_180(bearing_test - ground_course_deg);
+
+        // don't consider 180 degree turns, as we can't predict which way we will turn
+        if (fabsf(bearing_change_deg) > 170) {
+            continue;
+        }
+
+        // ToDo: add effective groundspeed calculations using airspeed
+        // ToDo: add prediction of vehicle's position change as part of turn to desired heading
+
+        // test location is projected from current location at test bearing
+        Location test_loc = current_loc;
+        test_loc.offset_bearing(bearing_test, lookahead_step1_dist);
+
+        // calculate margin from fence for this scenario
+        float margin = calc_avoidance_margin(current_loc, test_loc);
+        if (margin > best_margin) {
+            best_margin_bearing = bearing_test;
+            best_margin = margin;
+        }
+        if (margin > _margin_max) {
+            // this bearing avoids obstacles out to the lookahead_step1_dist
```

Comment:
Comment is incorrect.  We should probably also have sanity checks that relate your `_margin_max` to `lookahead_step1_dist`

Reply:
Is the comment incorrect?  In any case, I've changed the following comment to, "// now check in there is a clear path in three directions towards the destination".
I've added sanity checking to the AP_OABendyRuler::set_config() call so we can be sure the _look_ahead distance is at least 1m and _margin_max is at least positive.  I've done something similar to AP_OADijkstra::set_fence_margin to ensure it gets a positive margin